Added learner details endpoint.

dsjen opened the pull request:
@mulby, @dan-f -- I've added the learner details endpoint. This PR incorporates elastic search and a stub for the learners list endpoint too.

There are a few entity and event fields that are a bit inconsistent between various specs and the real data, but they should be easy to update when we get aligned on the names.

Also, the path has updated here instead of /api/learner_analytics/v0/learners, it is now /api/v0/learners for both consistency and potential changes to the API recommendation.

dan-f commented on analytics_data_api/v0/serializers.py:
```diff
@@ -306,3 +310,30 @@ class Meta(object):
+        for entity_type in engagement_entity_types.ALL:
+            for event in engagement_events.EVENTS[entity_type]:
+                metric = '{0}_{1}'.format(entity_type, event)
+                print metric
```
Forgot to remove a couple prints in here

dsjen replied:
Thanks!  Updated.

dan-f commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -0,0 +1,113 @@
+    Captures the course_id query arg and validates it.
+    """
+
+    COURSE_ID_PATTERN = r'(?P<course_id>[^/+]+[/+][^/+]+[/+][^/]+)'
```
This is also defined in `analytics_data_api/v0/urls/courses.py`. Could we share it?

dan-f commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -0,0 +1,113 @@
+        return super(CourseViewMixin, self).get(request, *args, **kwargs)
+
+
+class LearnersListView(CourseViewMixin, generics.ListAPIView):
```
This is part of the next story, right?

dsjen replied:
Oops!  I'll remove the learn list view and route.

dan-f commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -0,0 +1,113 @@
+    course_id = None
+
+    def get(self, request, *args, **kwargs):
+        self.course_id = request.QUERY_PARAMS.get('course_id', None)
```
We should use the `opaque_keys` library to validate the `course_id`.

dsjen replied:
Good idea!  Updated.

dan-f commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -0,0 +1,113 @@
+        queryset = self.get_queryset()
+        if len(queryset) == 1:
+            return queryset[0]
+        raise Http404("Object does not exist")
```
Check out the error format [here](https://openedx.atlassian.net/wiki/display/AC/edX+REST+API+Conventions#edXRESTAPIConventions-5.Errors) - we should make sure we're following that. I'm not sure how Http404 is treated within DRF.

dsjen replied:
Thanks, I hadn't seen that before.  Do you have an example of how that's implemented in the LMS?

dan-f commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,90 @@
+        self.assertDictEqual(expected, response.data)
+
+    @patch('analytics_data_api.v0.models.RosterEntry.get_course_user', Mock(return_value=[]))
+    def test_not_found(self):
```
404s happen when the username/course_id pair aren't found. We should try this with both an invalid username and an invalid course_id. Perhaps use `ddt` here?

dsjen replied:
Do you mean that we throw a 404 when the username and course_id is malformed or simply that it they don't exist?

dsjen replied:
Thanks for the clarification.  I wasn't sure if by "invalid" you meant malformed (e.g. course ID of "not-a/course").  For malformed, we throw a 400.

dan-f commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,90 @@
+            "email": "ed_xavier@example.com",
+            "segments": ["has_potential"],
+            "engagement": {
+                "forum_commented": 0,
```
These aren't the keys in the API doc.

dsjen replied:
I pulled these out from the the "possible events" in https://openedx.atlassian.net/wiki/display/AN/Learner+Analytics.  Also, the naming isn't quite consistent between what's in the documentation, api spec, and actual index though.  On the API: spec we have:
* discussions_contributed
* problems_attempted
* problems_completed
* videos_viewed
* videos_completed

The table spec has:
* problem_attempted
* problem_completed
* video_played
* forum_created
* forum_responded
* forum_commented

The current index payload has:
* problems_attempted
* problems_completed
* discussion_activity
* videos_watched

@dan-f & @mulby -- does anyone have opinions on what the names should be?

dan-f replied:
A recap from our huddle. Event names should be as follows:

- problem_attempts
- problems_attempted
- problems_completed
- videos_viewed
- discussions_contributed

dsjen replied:
Thank you!  Updated.

dan-f commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,90 @@
+            "name": "Edward Xavier",
+            "email": "ed_xavier@example.com",
+            "segments": ["has_potential"],
+            "engagement": {
```
The API doc chooses they key "engagements" rather than "engagement". I don't really mind either way, but we should be consistent.

dsjen replied:
Ah!  I'll update to "engagements" instead.

dan-f commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,90 @@
+        self.assertEquals(response.status_code, 200)
+
+        expected = {
+            "username": "ed_xavier",
```
We need to add "account_url"

dsjen replied:
Added LMS_USER_ACCOUNT_BASE_URL to settings and updated serializer to return account_url field.

dan-f replied:
Great!

dan-f commented:
Done with my first pass, @dsjen. I'm still not convinced about the old `api/v0/...` url structure, as it boxes us in. I looked a bit at the URL routing code. Due to the separation of the `analyticsserver` and `analytics_data_api` apps, we should be able to include another `learner_analytics` app in the same way that we include the `analytics_data_api` or `v0` apps.

https://github.com/edx/edx-analytics-data-api/blob/master/analyticsdataserver/urls.py#L15
https://github.com/edx/edx-analytics-data-api/blob/master/analytics_data_api/urls.py#L6

dsjen commented:
Thanks, @dan-f!  I had the URLs routed and had to add a bit of extra code at https://github.com/edx/edx-analytics-data-api/compare/dsjen/learner-engagement-timeline?expand=1#diff-1b6a5ff34357efc621dd247bb536cb3eR4, https://github.com/edx/edx-analytics-data-api/compare/dsjen/learner-engagement-timeline?expand=1#diff-110b488f2b07b14fdb4a2589e7da9e18R9, and https://github.com/edx/edx-analytics-data-api/compare/dsjen/learner-engagement-timeline?expand=1#diff-92cd74a9de7813a9d844d93410cf5cf3R198.  There's a bit of throw away code involved because we're supporting two conventions.

I wanted to provide a few links for reference and lets chat more.  I'll address your other comments in the meantime.

mulby commented on analytics_data_api/v0/models.py:
```diff
@@ -206,3 +208,22 @@ class Video(BaseVideo):
+
+
+if settings.ELASTICSEARCH_LEARNERS_HOST:
+    connections.connections.create_connection(hosts=[settings.ELASTICSEARCH_LEARNERS_HOST])
```
I think we should avoid executing logic like this on import.

I feel like this should really go in a "one-time startup" function like the one described here: https://docs.djangoproject.com/en/1.8/ref/applications/#django.apps.AppConfig.ready

I know we hacked together a "startup" function for the LMS, but since then I think Django has added first class support for that type of pattern. I assume it's the method I linked to above, but am not really sure, maybe worth talking to someone who is familiar with the newer versions of Django.

dan-f replied:
:+1: to this. Could we alternatively add it to the django settings module?

dsjen replied:
Thanks, all!  I went with the `AppConfig.ready` method--seems to work well.

mulby commented on analytics_data_api/v0/models.py:
```diff
@@ -206,3 +208,22 @@ class Video(BaseVideo):
+if settings.ELASTICSEARCH_LEARNERS_HOST:
+    connections.connections.create_connection(hosts=[settings.ELASTICSEARCH_LEARNERS_HOST])
+
+# By creating the roster index, we can use the @roster.doc_type decorator below
```
This comment implies that the index is actually created here, which I think is a little confusing.

dsjen replied:
Makes sense.  I'll update the comment.

mulby commented on analytics_data_api/v0/models.py:
```diff
@@ -206,3 +208,22 @@ class Video(BaseVideo):
+
+    @classmethod
+    def get_course_user(cls, course_id, username):
+        return cls.search().filter('term', course_id=course_id).query(
```
Not too familiar with the syntax here, but I profiled the following query:

```json
{
    "size": 50,
    "from": 0,
    "query": {
        "bool": {
            "must": [
                {
                    "term": {
                        "roster_entry.username": "foo"
                    }
                },
                {
                    "term": {
                        "roster_entry.course_id": "bar"
                    }
                }
            ]
        }
    }
}
```

dsjen replied:
That's interesting.  I would have expected the course ID to be wrapped in a `filter` and the username to nested in a `match` field.  As I understand, the filter component is a caching optimization.  I'm going to punt on that for now and just set these to be both `query` and `term` and the result should be the same.

mulby commented on analytics_data_api/v0/models.py:
```diff
@@ -206,3 +208,22 @@ class Video(BaseVideo):
+
+
+@roster.doc_type
+class RosterEntry(DocType):
```
I'm a little surprised you don't have to declare the fields here. Are there pros and cons associated with declaring them or not?

Part of me likes the explicitness of declaring them.

mulby commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,91 @@
+            },
+        }
+        from pprint import pprint
+        pprint(response.data)
```
don't forget to remove this!

dsjen replied:
Oops!  I'll remove that.

mulby commented:
Finished my pass, let me know when the URLs are re-configured and I can take another look.

dan-f commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -0,0 +1,88 @@
+        try:
+            CourseKey.from_string(self.course_id)
+        except InvalidKeyError:
+            raise ParseError('Course ID malformed')
```
I'm curious as to what code path is triggered by this `ParseError`. Does it result in an HTTP 400? If so, what's the message?

dsjen replied:
Yes, this returns a 400, but not in the edx recommended format.  I'll see about updating that.

dan-f commented on analyticsdataserver/settings/base.py:
```diff
@@ -271,7 +271,11 @@
 
 ENABLE_ADMIN_SITE = False
 
+# base url to generate link to learner profile page
```
This should actually be the URL to generate a link to the User API, not the HTML page.

dsjen replied:
Ah, I see!  I'll update the comment.

dan-f commented:
Let me know when you get to error messages stuff. Looking good!

dsjen commented:
@dan-f & @mulby -- I believe I've addressed all comments now.  Of note is that I've added custom exceptions and middlewares to catch them, which return error codes and message.

dan-f commented on analytics_data_api/v0/apps.py:
```diff
@@ -0,0 +1,13 @@
+    def ready(self):
+        super(ApiAppConfig, self).ready()
+        if settings.ELASTICSEARCH_LEARNERS_HOST:
+            connections.connections.create_connection(hosts=[settings.ELASTICSEARCH_LEARNERS_HOST])
```
[This page](https://docs.djangoproject.com/en/1.9/ref/applications/#django.apps.AppConfig.ready) warns against interacting with the database in the `ready` method. Apparently it's called every time a management command is invoked. Since we're not querying Elasticsearch here, it's probably not an issue, right? cc @mulby

dsjen replied:
Good point!  We're mocking out the calls in our unit tests, so this shouldn't be hit.

dan-f commented on analytics_data_api/v0/exceptions.py:
```diff
@@ -0,0 +1,52 @@
+    Base error.
+    """
+
+    __metaclass__ = abc.ABCMeta
```
Why do we need `abc`? I thought that the `Exception` class provides this behavior on its own.

dsjen replied:
I'm preventing others from creating an instance of `BaseError` directly and use the other concrete classes defined in the file.  I believe without `abc` once could still create an instance of an `BaseError` otherwise.

dan-f replied:
I said it poorly, but I think what I meant was that the abc/message/__str__ stuff may be unnecessary because the base Exception class provides that functionality itself. I don't mind leaving this, though.

dan-f commented on analytics_data_api/v0/tests/views/test_learners.py:
```diff
@@ -0,0 +1,88 @@
+        base_path = '/api/v0/learners/{}'
+        path = (base_path).format('ed_xavier')
+        response = self.authenticated_get(path)
+        self.assertEquals(response.status_code, 400)
```
One final request: I'd prefer if the test cases are split up by edge cases rather than error codes. It makes the test cases more targeted and descriptive; we could have a `test_no_course_id`, and a `test_bad_course_id` rather than `test_get_400`, for example.

Also, now that we have awesome error handling, could you verify that the JSON corresponding to the error codes is correct?

dan-f commented:
Looks great to me, @dsjen. Thanks for being responsive to feedback! I love the use of middleware for error handling. :+1: once you address my two questions and one suggestion.

dsjen commented:
@dan-f -- thanks!  I've separated the tests out into two methods that verify the json payload against expected results.
